mixpanel_oauth_to_redshift.go

```diff
@@ -0,0 +1,177 @@
+	redshift_table       = flag.String("redshift_table", "mixpanel_oauth", "Name of the redshift table.")
+	export_from_mixpanel = flag.Bool("export_from_mixpanel", true, "Whether to export from mixpanel.")
+	copy_to_redshift     = flag.Bool("copy_to_redshift", true, "Whether to copy to redshift.")
+)
+
+type MixpanelExport struct {
+	endpoint   string
+	version    string
+	api_key    string
+	api_secret string
+}
+
+func NewMixpanelExport() *MixpanelExport {
+	m := MixpanelExport{"https://data.mixpanel.com/api", "2.0", mixpanel_api_key, mixpanel_api_secret}
+	return &m
+}
+
+// json.Marshal wraps a string with quotes which breaks URL params in API calls. This function fixes
+// that problem.
+func StringOrJsonMarshal(value interface{}) (string, error) {
+	str_value, ok := value.(string)
+	if !ok {
+		json_value, err := json.Marshal(value)
+		if err != nil {
+			return "", err
+		}
+		str_value = string(json_value)
+	}
+	return str_value, nil
+}
+
+func (m *MixpanelExport) ComputeSig(params map[string]interface{}) (string, error) {
+	var keys []string
+
+	for k := range params {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	joined_args := ""
+	for _, k := range keys {
+		joined_args += k
+		value, err := StringOrJsonMarshal(params[k])
+		if err != nil {
+			return "", err
+		}
+		joined_args += fmt.Sprintf("=%s", value)
+	}
+
+	hash := md5.Sum([]byte(joined_args + m.api_secret))
+	return hex.EncodeToString(hash[:]), nil
+}
+
+func (m *MixpanelExport) Request(method string, params map[string]interface{}) ([]byte, error) {
+	params["api_key"] = m.api_key
+	expire_time := time.Now().Add(*mixpanel_export_request_timeout)
+	params["expire"] = expire_time.Unix()
+	params["format"] = "json"
+	delete(params, "sig")
+	sig, err := m.ComputeSig(params)
+	if err != nil {
+		return nil, err
+	}
+	params["sig"] = sig
+
+	values := url.Values{}
+	for key, value := range params {
+		str_value, err := StringOrJsonMarshal(value)
+		if err != nil {
+			return nil, err
+		}
+		values.Set(key, str_value)
+	}
+	url := fmt.Sprintf("%s/%s/%s/?%s", m.endpoint, m.version, method, values.Encode())
+	log.Println(url)
+	resp, err := http.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	return ioutil.ReadAll(resp.Body)
+}
+
+func CopyJsonDataFromS3ToRedshift(
+	db *sql.DB, redshift_table string, export_file string, aws_access_key_id string,
+	aws_secret_access_key string, jsonpaths_file string, aws_region string) error {
+	copy_cmd := fmt.Sprint(
+		"COPY ", redshift_table, " FROM '", export_file, "' WITH ",
+		"credentials ", fmt.Sprintf("'%s=%s;%s=%s'", "aws_access_key_id", aws_access_key_id, "aws_secret_access_key", aws_secret_access_key),
+		" json '", jsonpaths_file,
+		"' region '", aws_region,
+		"' timeformat 'epochsecs'")
+	log.Print("Executing command: ", copy_cmd)
+	_, err := db.Exec(copy_cmd)
+	return err
+}
+
+func main() {
+	flag.Parse()
+	export_file := fmt.Sprintf("%s/%s", *mixpanel_export_dir, *mixpanel_export_date)
```

Comment:
I tried that but strips out one of the `/` from the s3 path breaking the filename.